sdks/go/pkg/beam/core/runtime/exec/datasource.go

```diff
@@ -173,10 +183,13 @@ func (n *DataSource) Process(ctx context.Context) error {
 }
 
 func (n *DataSource) FinishBundle(ctx context.Context) error {
-	log.Infof(ctx, "DataSource: %d elements in %d ns", atomic.LoadInt64(&n.count), time.Now().Sub(n.start))
+	n.mu.Lock()
+	log.Infof(ctx, "DataSource: %d elements in %d ns", n.count, time.Now().Sub(n.start))
 	n.source = nil
 	err := n.Out.FinishBundle(ctx)
-	atomic.StoreInt64(&n.count, 0)
+	n.count = 0
+	n.splitPos = math.MaxInt64
+	defer n.mu.Unlock()
```

Comment:
The defer needs to be immeadiately after the n.mu.Lock() call to be effective. If FinishBundle panics, this line would never be reached.

defer can be read as  "no matter how this FinishBundle method exits, ensure this code runs." Kind of like Java's "finally" in a "try {} catch {}finally {}" paradigm.

Reply:
done!